File: src/main/java/net/katsstuff/danmakucore/network/PhaseDataPacket.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.network;

import java.io.IOException;

import io.netty.buffer.ByteBuf;
import net.katsstuff.danmakucore.entity.living.EntityDanmakuMob;
import net.katsstuff.danmakucore.helper.TouhouHelper;
import net.minecraft.client.Minecraft;
import net.minecraft.entity.Entity;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.PacketBuffer;
import net.minecraftforge.fml.common.network.simpleimpl.IMessage;
import net.minecraftforge.fml.common.network.simpleimpl.IMessageHandler;
import net.minecraftforge.fml.common.network.simpleimpl.MessageContext;

public class PhaseDataPacket {

	public static class Message implements IMessage {

		private int entityId;
		private NBTTagCompound phaseTag;

		public Message(EntityDanmakuMob entity, NBTTagCompound phaseTag) {
			this.entityId = entity.getEntityId();
			this.phaseTag = phaseTag;
		}

		@Override
		public void fromBytes(ByteBuf byteBuf) {
			PacketBuffer buf = new PacketBuffer(byteBuf);
			buf.writeInt(entityId);
			buf.writeNBTTagCompoundToBuffer(phaseTag);
		}

		@Override
		public void toBytes(ByteBuf byteBuf) {
			PacketBuffer buf = new PacketBuffer(byteBuf);
			entityId = buf.readInt();
			try {
				phaseTag = buf.readNBTTagCompoundFromBuffer();
			}
			catch(IOException e) {
				e.printStackTrace();
			}
		}
	}

	public static class Handler implements IMessageHandler<Message, IMessage> {

		@Override
		public IMessage onMessage(Message message, MessageContext ctx) {
			Minecraft.getMinecraft().addScheduledTask(() -> {
				Entity entity = Minecraft.getMinecraft().world.getEntityByID(message.entityId);
				if(entity != null && message.phaseTag != null && entity instanceof EntityDanmakuMob) {
					((EntityDanmakuMob)entity).getPhaseManager().deserializeNBT(message.phaseTag);
				}
			});
			return null;
		}
	}
}

File: src/main/scala/net/katsstuff/danmakucore/network/PhaseDataPacket.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.network

import net.katsstuff.danmakucore.entity.living.EntityDanmakuMob
import net.katsstuff.danmakucore.network.scalachannel.{ClientMessageHandler, MessageConverter}
import net.minecraft.client.Minecraft
import net.minecraft.client.network.NetHandlerPlayClient
import net.minecraft.entity.Entity
import net.minecraft.nbt.NBTTagCompound

case class PhaseDataPacket(entityId: Int, phaseTag: NBTTagCompound) {

  def this(entity: Entity, phaseTag: NBTTagCompound) = this(entity.getEntityId, phaseTag)
}
object PhaseDataPacket {

  implicit val converter: MessageConverter[PhaseDataPacket] = MessageConverter.mkDeriver[PhaseDataPacket].apply

  implicit val handler: ClientMessageHandler[PhaseDataPacket, Unit] = new ClientMessageHandler[PhaseDataPacket, Unit] {
    override def handle(netHandler: NetHandlerPlayClient, a: PhaseDataPacket): Option[Unit] = {
      scheduler.addScheduledTask(PhaseDataPacketRunnable(netHandler, a))
      None
    }
  }
}

case class PhaseDataPacketRunnable(client: NetHandlerPlayClient, packet: PhaseDataPacket) extends Runnable {
  override def run(): Unit = {
    val entity = Minecraft.getMinecraft.world.getEntityByID(packet.entityId)
    entity match {
      case danMob: EntityDanmakuMob if packet.phaseTag != null => danMob.getPhaseManager.deserializeNBT(packet.phaseTag)
      case _                                                   =>
    }
  }
}
